### ISSUE TITLE
Depth/Nesting member/method

### ISSUE THREAD

**Issue by wbuchanan** [FasterXML/jackson-future-ideas#7]
Is it possible to have some type of int member that would indicate the depth of the nesting for specific nodes/arrays and/or the depth of the most nested element in a given node?  

```
                                    great-grandparent                                   depth = 0
              grandparent                                     grandparent               depth = 1
         parent                    parent                 parent            parent      depth = 2
child          child                                   child              child child   depth = 3
               newborn                        newborn newborn                           depth = 4
```

The value of depth could be used to create unambiguous replacement keys within their lineage (e.g., there would be a newborn1 and newborn2 for the third child element in the example above).  A maximum depth value for a given node would also make it easier to rectangularize the data by providing the number of times a given key/value pair might need to be repeated.

**Comment by cowtowncoder** [comment]
It is possible to calculate this from read/write context, traversing parents. As to adding explicit tracking, I don't think I would personally be interested in adding such information; whoever calls parser/generator could/should keep track of that.

**Comment by wbuchanan** [comment]
@cowtowncoder perhaps adding other information related to the child elements could help?  For example, including a hasChildren() method could be useful as a stopping/continuation rule for these types of tasks (particularly in the context where the JSON being parsed is arbitrary and could have any number of nested elements).  It could also prevent having to read over the data multiple times to obtain information that I suspect is easier to track/compute during the construction of a JsonNode/TreeNode object.

**Comment by cowtowncoder** [comment]
@wbuchanan including `hasChildren()` where? Note that `JsonStreamContext`, exposed by both `JsonParser` and `JsonGenerator` does have method `getEntryCount()` that does give indication as to how many preceding siblings there has been.

**Comment by wbuchanan** [comment]
@cowtowncoder Sorry about that...was a bit early in the morning when I was writing that.  I was thinking in the different classes in the databind.node package.  I also threw together some stuff that could be used to return either the maximum depth of all elements in a given node, or would return the list of depths for each of the child nodes [here](https://github.com/wbuchanan/StataJSON/blob/flatTest/src/main/java/org/paces/Stata/Ingest/JsonUtils.java).  

Basically I'm trying to find ways that I can reconstruct the JSON in a two dimensional model (e.g., a normalized/denormalized SQL table).

**Comment by cowtowncoder** [comment]
@wbuchanan Ok so this would be for tree model, exclusively, not for streaming parser. It is possible to actually implement custom tree models; either by sub-classing various `JsonNode` types (and `JsonNodeFactory`), or by just implementing `TreeNode` and `TreeCodec`. But not sure if that would simplify task signifciantly.

**Comment by wbuchanan** [comment]
@cowtowncoder yea.  So instead of just having methods that return the current node type, there are methods to check the types of the children and those get used to determine whether there are any child elements that need further descent.  I don't personally have a ton of experience with Java, so I'm also not sure sub-classing or implementing the interfaces would make things easier in the immediate future (although I imagine it would be the better or at least more proper approach).

**Comment by cowtowncoder** [comment]
At this point I don't think there is value in explicitly counting nesting depth; I have not seen use cases nor had anyone requesting it for solving existing problem. So closing for now.

**Comment by andrewfinnell** [comment]
Perhaps there is a better way to track this information, but I am in need of something like this as well.

If you have a Container which can contain other Container's, it's impossible to only serialize the first level of Containers. It's also impossible to limit the Depth by a Scalar number. As the Serializers are set based on Class type, you cannot differentiate between an Item which contains other Items, and an Item that is Container by another Item, in order not to recursively serialize that Item's contents.

**Comment by cowtowncoder** [comment]
There is one way to keep track of contextual information, and that is via "Attributes" accessed through `SerializerProvider` context object. Serializers can access it. But that is not something generator object can access, only databind entities lke `JsonSerializer`. Serializers themselves can not track state as they are not contextual or thread-safe (that is, one serializer may be concurrently be used by multiple operations).

Conversely `JsonGenerator` does also have something applicable:

* `getOutputContext()` returns information on current logical position (are we withing array, object, or root level) and nesting (may access parent context)
* Context (individual level) also has optional "current value", set by `JsonSerializer`

combination of which does allow some amount of traversal about parent hierarchy (nesting).

Now, as to limiting nesting: if I was to work on that, I would rather spend time on serializers tracking nesting and limits, and not trying to provide general-purpose depth counters. Reason here being that keeping track of information needed for enforcing limits is easier and cheaper to do knowing what limits are needed and when -- general-purpose counters are unlikely to be good for specific use cases.

Or put another way: if the goal is to support limitations on traversal during serialization, it is better to start with that, instead of requesting general depth counters. Even if counters are chosen as the mechanism, it is good to start with problem to solve, to consider different possible routes.
